### DISCUSSION TITLE
Tidying up misidentifications

### DISCUSSION THREAD

**Discussion by srchild** [mcguirepr89/BirdNET-Pi#262] [Q&A] [answered]
I intend making my installation public, but it will look silly if I leave significant misidentifications uncorrected (e.g. birds which are not known in this country)

To delete a misidentified clip/bird, is it sufficient to delete the relevant row(s) from the database. Is there anything else to do, such as a reindex after manipulating the database?

If I leave the audio files in place will they eventually get deleted automatically, or should I delete the audio file as well as the database row to avoid wasting SD card space on redundant audio files?

Thanks

**Comment by mcguirepr89** [comment]
>To delete a misidentified clip/bird, is it sufficient to delete the relevant row(s) from the database. Is there anything else to do, such as a reindex after manipulating the database?

Deleting from the database will be sufficient to remove the data from the "Species Stats" page, and from the "Today's Detections," "Best Recordings," and "Recordings" pages.

>If I leave the audio files in place will they eventually get deleted automatically, or should I delete the audio file as well as the database row to avoid wasting SD card space on redundant audio files?

To actually remove the audio files, I made a little bash alias that can be invoked as `remove` and then some really unique identifier. For instance, to remove all of the "Common Nighthawk" recordings, but keep the "Red-shouldered Hawk" recordings, I would issue "`remove "Common Nighthawk"`" in the terminal and that should work. If you have a species with a really unique name, you can use the unique part. For instance, removing the "Swainson's Warbler" could be accomplished with just "`remove swainson`" as it is unique enough to only catch the species I need to remove.

Use that tool carefully, as it does not ask before removing files.

Another method to remove the undesirable audio clips could be:
```
find ~/BirdSongs -type d -name '*Common Nighthawk*' -exec rm -drfv {} +
```

Lastly, to prevent those species from showing up in your database again in the future, you can prevent BirdNET-Pi from reporting predictions for those species by adding them to your "Excluded Species List".

To do this, go to "Tools" > "Excluded Species List"

On that page, press "Ctrl + F" to enter your browser's search mode. Type the name of the species you wish to exclude from your reporting and data collection, select from within the list of species in the box on the left and click ">>ADD>>" to have the species added to your Excluded Species List.

Please let me know if you have any other questions and I'll do my best to help out.

My very best regards,
Patrick

**Reply by srchild** [reply]
Thanks for that prompt help

**Reply by CaiusX** [reply]
How about an option to move misidentified sounds to a misidentified directory?  Could be useful for future model training.

**Reply by Tankeboks** [reply]
I have had a few misidentifications (it tends to identify my drill as some bird that is extinct here) .. would it not make sense to put a button on the specific recording that lets you mark something as a false positive that will then also tidy up and remove the clip + the database entry ? (or just add a false positive = 1 to the database entry?)

**Reply by srchild** [reply]
I have had just the same thought but hesitated to make yet another change request!

Yes, an option / checkbox / something that can be selected/pressed when viewing a recording, which will delete both the recording and the corresponding database entry in a single action, would be very helpful. I spent some time a few days ago checking suspect recordings, finding them in the db and deleting them, and then finding the actual recordings and deleting them.

It would be useful not only for misidentifications but also for recordings that need deleting for other reasons such as human speech being recorded.

**Comment by jds3d** [comment]
Are there instructions somewhere for connecting to the database/deleting the rows?  I have an eastern screech owl record that I played from Merlin to test the original setup.  I can write the sql if I have the connections details and info on what tables need to be cleared.

**Reply by mcguirepr89** [reply]
Hi, @jds3d -- I imagine you'll be able to intuit the proper steps after navigating to "Tools" > "Database Maintenance."

There, you will be able to do anything you need to the database. Please consult the [Adminer website](https://www.adminer.org/) for more information.

Alternatively, if you prefer using SQL queries:
```
sqlite3 ~/BirdNET-Pi/scripts/birds.db
```
will get you to an SQL prompt. The table is `detections`

Let me know if you have any trouble ~~

**Reply by jds3d** [reply]
Ah, I see the issue.  I am unable to connect to tools.  When I click it, it gives me a login prompt, but the login details to the raspberry pi do not work and I don't have another login to try.  I will try the sqlite3 in the mean time.

**Reply by jds3d** [reply]
For posterity, after 'ssh'ing in, it only took two commands to remove the detections of Eastern Screech-Owl.

```
sqlite3 ~/BirdNET-Pi/scripts/birds.db
delete from detections where com_name = 'Eastern Screech-Owl';

```

**Reply by mcguirepr89** [reply]
Hi, @jds3d -- _please_ read the [main page's Access section](https://github.com/mcguirepr89/BirdNET-Pi#access) (as well as the rest of the ReadMe and Wiki) to get familiar with the system. If you have trouble locating documentation or find it lacking, please let us know.

**Reply by slowrunner** [reply]
@mcguirepr89 >If you have trouble locating documentation or find it lacking, please let us know.

The docs for getting installed and configured were great.  The very first question I had that I could not find in the README, nor Wiki, nor birdnetpi.com, nor in the open or closed issues was:

"What do I do about the Northern Mockingbird (Florida state bird) showing up all day as black-whiskered vireo, tufted titmouse, and Baltimore oriole?"

Followed next by "How do I delete the Eastern Screech Owl detection from a passing ambulance?"

BTW, some of the misidentifications were higher confidence than the correct Northern Mockingbird detections.

My last question was to find where BirdNet-pi users are helping each other and asking questions - which it appears the Git Discussions page is serving well - I need to continue reading all these discussions.

The installation of birdnet-pi code was the coolest, flawless install I have ever seen - with auto config of the Linux sound system to find and use my USB mic was simply magical.  Do you have a kofi or other tip-jar?

**Reply by slowrunner** [reply]
@jds3d Thank you for the sql to remove unwanted detections - worked like a charm my first day collecting.

Would you be so kind as to offer the SQL for deleting a species that I want to qualify the where clause with "date equals today"?

**Reply by mcguirepr89** [reply]
Hi, @slowrunner -- thanks for your kind words about the project.

Here are two things I hope are helpful for your database maintenance. 

1) The Database Maintenance tool in "Tools" can be really helpful in achieving its namesake. 
   In my example below, I will use the Database Maintenance tool to remove all instances of the "Common Raven" from 2022-05-25
   
   Login (no password, just click the button)
   ![image](https://user-images.githubusercontent.com/60325264/176191857-a8528d25-53b7-41dc-b618-c8efcd20c9c2.png)
   Select the "select" button for the "detections" table using the navigation menu on the left
   ![image](https://user-images.githubusercontent.com/60325264/176191938-c3f67857-8938-432e-9b2c-287e45feccad.png)
   On the page that displays the "detections" table, click the second option in the top row of tools to open the "Search" tool
   ![image](https://user-images.githubusercontent.com/60325264/176191978-63ac5d96-12fa-4a4d-b208-d1cfc20ab17e.png)
   Choose the field you are interested in -- for our purposes, we select the "Com_Name" with the "LIKE%%" operator (this will search for any match, case insensitive) and "Date" with the "=" equals operator to only look for entries from 2022-05-25. Click "Select" on the right to filter the results.
   ![image](https://user-images.githubusercontent.com/60325264/176192142-cea106af-4e55-4802-8486-742d83281c56.png)
   To delete all of the returned results, click the "Whole result" box on the bottom left and then the "Delete" button to its right
   ![image](https://user-images.githubusercontent.com/60325264/176192180-db7a41cb-9a08-4f41-a037-1c3d8fcca5c6.png)
   Then those entries are gone from the database
   ![image](https://user-images.githubusercontent.com/60325264/176192242-a34e80d7-a1ec-41f4-a420-8717210b5681.png)

2) To perform the exact same steps above with a single SQL command, you can simply add an "AND" clause to your "WHERE" clause in your sql statement:
   ```
   DELETE FROM detections WHERE Com_Name LIKE '%raven%' AND Date = "2022-05-25"
   ```
   or
   ```
   DELETE FROM detections WHERE Com_Name = "Common Raven" AND Date = "2022-05-25"
   ```
   or, to address your request that the clause mean "where date equals today"
   ```
   DELETE FROM detections WHERE Com_Name = "Species Common Name" AND Date = DATE('now', 'localtime')
   ```
I hope that is helpful for you -- I will do my best to improve the documentation

**Reply by slowrunner** [reply]
Superb.  This works to delete mis-identifications.

There is a panel that seems for the purpose of manually identifying/correcting mis-id'd birds under species stats, that appears to offer a way to change misidentifications.

Under Species Stats:
1) select a species
2) listen to recording
3) click reviewed
4) Select "False Positive"
5) Edit/selection box with "UNKNOWN" appears, type Norther M, select Northern Mockingbird

But then what?  It lacks a "Save New ID" button.

**Reply by slowrunner** [reply]
Using the Database Maintenance concept, I was able to edit the com_name and sci_name of the misidentified records, and see the effect when I checked species stats again.  I can do the "tidying up misidentifications" this way also.

**Reply by CaiusX** [reply]
Hi y’all

will push this facility thru sooner than later. It’s working on joburg.birdnetpi.com under species stats page, African electricity and GitHub ignorance are retarding the workflow at the moment.

**Reply by FirstTenor** [reply]
> Superb. This works to delete mis-identifications.
> 
> There is a panel that seems for the purpose of manually identifying/correcting mis-id'd birds under species stats, that appears to offer a way to change misidentifications.
> 
> Under Species Stats:
> 
> 1. select a species
> 2. listen to recording
> 3. click reviewed
> 4. Select "False Positive"
> 5. Edit/selection box with "UNKNOWN" appears, type Norther M, select Northern Mockingbird
> 
> But then what? It lacks a "Save New ID" button.

I saw this too …. I have successfully used the functions to ‘delete re incorrect recording’ - but would really prefer to someone flag / report these misidentifications to improve the model.  This function that @slowrunner quotes seems partially implemented… as mentioned, you can edit ‘Unknown’ to provide the real answer - but not save it.

I also wonder if that would just be ‘my local instance’ or actually reported back somewhere centrally?

Current examples (during daylight) are:

Tawny Owl = Police Car 
Small Eurasian owl = Rook & Robin calling concurrently
Small toed treecreeper = someone walking dog in woods over mains hum

**Reply by danny7989** [reply]
I, too, was wondering about this review feature. I have false positives which I would like to correct/review, and this feature would be ideal if it worked. If the review influenced the AI, it would be even better. Reviewing or deleting false positives from the DB directly is not currently an option for me.

**Comment by tomwhite101** [comment]
I would be great to be able to delete a record directly from the species list.

**Reply by mcguirepr89** [reply]
I agree -- this will be possible eventually, hopefully sooner than later

**Comment by GeekNJ** [comment]
Just getting started and there was a major misidentification. We could delete the record or maybe just a field to identify as incorrect in the DB if there's ever a reason to use that data in the future for the project. Just a thought.

**Comment by hoover67** [comment]
I've deleted many misidentifications using the adminer tool, but it's a bit cumbersome. A link to delete the spectrogram, sound file and db entry for an MI would be great and I'd be happy to help implementing it (I'm a django / python guy though most of the time when it comes to web stuff). 

All the best, 

Uwe

**Comment by hoover67** [comment]
`http://birdnetpi.local/scripts/adminer.php?sqlite=&username=&db=birds.db&select=detections&columns%5B0%5D%5Bfun%5D=&columns%5B0%5D%5Bcol%5D=&where%5B0%5D%5Bcol%5D=&where%5B0%5D%5Bop%5D=LIKE+%25%25&where%5B0%5D%5Bval%5D=sandp&where%5B01%5D%5Bcol%5D=&where%5B01%5D%5Bop%5D=%3D&where%5B01%5D%5Bval%5D=&order%5B0%5D=&limit=50&text_length=100`

Here's a quick URL for your bookmarks which will take you straight to the adminer interface and offering a LIKE searchfield. Not exactly what we're asking for but it saves a few clicks. 

Cheers, Uwe

**Reply by ehpersonal38** [reply]
Hi, I just implemented something that will hopefully meet your needs. Go ahead and do an update and then you should be able to delete detections from the DB by either 1) clicking the copy link in the top right of a detection on Overview or Today's Detections, then clicking X in the top right of the page that opens or 2) from the Recordings page. Let me know how that goes for you.

Ethan

**Reply by mcguirepr89** [reply]
Thanks for this, Ethan!! Super helpful

**Reply by mcguirepr89** [reply]
Not sure how to mark that as the answer?

**Reply by Phaeton** [reply]
You should be able to uncheck the prior one. BUT, the answer cannot be a reply. Ethan @ehpersonal38 would need to resubmit as an answer.

**Comment by ehpersonal38** [comment, marked as answer]
d2d914f8ee783ff466cb76cd40e950896e937202 offers a way to natively delete detections now in the interface! Hopefully this serves as a stepping stone to build out things further. I know there was some work done on addressing misidentifications in the `verification` branch for potentially retraining the model, interested to see how that progresses.

**Reply by mcguirepr89** [reply]
It (the `verification` branch) needs some tweaks to the workflow -- this is a great addition to it for sure!!!

**Reply by Sp4rkR4t** [reply]
This is a great addition but is there a reason the delete detection button is only in the recodings section, it would be superb if it was also available in todays detections section as that is the section I use most often to browse the days activity (it's only a single click vs 3 for that day in the recordings section)

**Reply by ehpersonal38** [reply]
I just did it that way since I think it would start looking a little crammed, and would probably look very bad on mobile. You can currently also delete detections this way though:

https://user-images.githubusercontent.com/103586016/179282574-68472df3-0230-4c98-adf1-55d20d705062.mp4

Does that work for you?

**Reply by GeekNJ** [reply]
An X usually represents closing vs deleting. I would suggest using a trash can to represent Delete.
